Feature: add billing address block

glaubersilva opened the pull request:
## Description

Depends on https://github.com/impress-org/givewp/pull/6837

<!-- Summarize the related issue, explain HOW this PR solves the problem, and WHY you made the choices you made. -->

This PR adds a billing address block that allows users to collect the donor's address during the donation process.

## Affects

<!-- Mention any existing functionality affected by this PR to help inform the reviewer(s). -->

Form Builder

## Visuals

<!-- Include screenshots or video to better communicate your changes. -->

![image](https://github.com/impress-org/givewp-next-gen/assets/16308864/5bb7616f-9cbd-444f-8f0e-a65762cc1daa)

![image](https://github.com/impress-org/givewp-next-gen/assets/16308864/084be756-3fce-4852-b421-3e80f56ac755)


## Testing Instructions

<!-- Help others test your PR as efficiently as possible. -->

1. Create a new block;
2. Add the new billing fields block to the form;
3. Save the form;
4. Create a new donation using the form;
5. Check the donations page to make sure that the billing address was saved.

## Pre-review Checklist

<!-- Complete tasks prior to requesting a review. Add to this list, but do not remove the base items. -->

-   [ ] Acceptance criteria satisfied and marked in related issue
-   [ ] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [ ] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed



---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205000286853460

jonwaldstein commented on src/DonationForms/DataTransferObjects/DonateFormRouteData.php:
```diff
@@ -83,8 +85,31 @@ public function validated(): DonateControllerData
+
         foreach ($validator->validated() as $fieldId => $value) {
-            $validData->{$fieldId} = $value;
+            switch ($fieldId) {
+                case 'country':
+                    $validData->billingAddress->country = $value;
+                    break;
+                case 'address1':
+                    $validData->billingAddress->address1 = $value;
+                    break;
+                case 'address2':
+                    $validData->billingAddress->address2 = $value;
+                    break;
+                case 'city':
+                    $validData->billingAddress->city = $value;
+                    break;
+                case 'state':
+                    $validData->billingAddress->state = $value;
+                    break;
+                case 'zip':
+                    $validData->billingAddress->zip = $value;
+                    break;
+                default:
+                    $validData->{$fieldId} = $value;
```
Ah I see why you had to do this.  It definitely works but adds additional granular complexity to this validation piece that I would like to avoid if possible.  

2 alternative ideas.
1. We assign all the billing address properties to the controller data DTO individually and add a new method called `getBillingAddress()` that builds the object when called.
2. We come up with a clever dot notion field name syntax that tries to auto-assign object properties.  Example:

```
$fieldName = 'billingAddress.country';
$field = explode('.', $fieldName);
$validData->{$field[0]}->{$field[1]} = $value;
```

I'm leaning toward option 1 since it's the simplest.  Open to your thoughts.

glaubersilva replied:
I liked the first option. Resolved in ec6cfbc54d3bb80fa7bf80c462305417b27a71d3

jonwaldstein commented on src/DonationForms/Actions/ConvertDonationFormBlocksToFieldsApi.php:
```diff
@@ -247,6 +252,51 @@ protected function createNodeFromDonorNameBlock(BlockModel $block): Node
+                ->placeholder($block->getAttribute('cityPlaceholder'))
+                ->rules('required', 'max:255');
+
+            $group->getNodeByName('state')
+                ->label($block->getAttribute('stateLabel'))
+                ->placeholder($block->getAttribute('statePlaceholder'))
+                ->rules('required', 'max:255');
```
We don't have to do this now, but I want to add a reminder that we'll probably need to make use of the visibility conditions for `state` so it's hidden for countries that don't have states.

glaubersilva replied:
I was taking a look at this and probably we'll need to make it respect not only the visibility according to the country, but also if it should be a text field or a select field. The v2 forms also load the states automatically on the front end when a select field is used. I recorded this video to demonstrate the different behaviors: https://www.loom.com/share/f530be8d1eca4ee682c7ee57e68203ce?sid=314a4195-2855-45a0-8dbe-19aee698f760

So, I believe that we need to handle it now because Ben already asked about it here:
https://www.loom.com/share/378a6f50a22e492babfb23371a0702da?highlightComment=12797555&t=66

jonwaldstein replied:
@glaubersilva interesting, yeah I guess we need to think about this one.  It also looks like GiveWP is loading all the state options on the server and looking for additional country/state settings saved.  I think we have a couple options for the new implementation:

1. Have the state field use the visibility conditions to register separate select & text fields with visibility and options based off the country (so only one works at at a time).  I'm not sure if this would work properly, but In theory this would mean we could accomplish everything with the fields api.
2. Make the state field a hidden field, and handle input changes and options on the client.

glaubersilva replied:
Resolved with a bunch of commits between Jun 28, 2023 and Jul 1, 2023

jonwaldstein reviewed (changes requested):
@glaubersilva great job on your first v3 form integration!  Just some minor comments.

Also I noticed in v2 forms we display the billing address in the confirmation page so we'll want to update that.  Here's a screenshot:

<img width="843" alt="Screenshot 2023-06-26 at 10 33 25 AM" src="https://github.com/impress-org/givewp-next-gen/assets/10138447/e3408a19-4bb4-4a9d-9667-d7e897107b9c">

That can be updated here (will need to update the matching unit test(s) from this update as well):

- `src/Framework/Receipts/Actions/GenerateConfirmationPageReceipt.php` 
- `fillDonorDetails()`

glaubersilva commented on src/Framework/FieldsAPI/BillingAddress.php:
```diff
@@ -0,0 +1,53 @@
+        $countryList = [];
+        foreach (give_get_country_list() as $value => $label) {
+            $countryList[] = [$value, $label];
+        }
```
Resolved in 491ac159c19ec56cd1341b89be64c70e104105cc and 7e4ad784f7a36e0411c9ea787cb5534738454e71

glaubersilva commented:
@jonwaldstein  Thanks for the suggestions, I implemented almost all of them. The donor details changes I implemented here d7abbe6c4fe22638b8f8340bac907a87d1e9f291

About the "state" field I left a comment with a few more details about how it works now. I believe that we should keep the same behavior from our existing billing address field.

But, before moving forward with the implementation, I would like to confirm if we should keep the same behavior from V2 forms or not.

glaubersilva commented:
@jonwaldstein I did some cosmetic changes to improve UI feedback and also change a few design details discussed in the comment of this video: https://www.loom.com/share/af7ad39687504710bb838b0b5eb3e5a7

So, it should be ready for a final review now.

jonwaldstein commented on src/FormBuilder/resources/js/form-builder/src/styles/_block-editor.scss:
```diff
@@ -116,6 +117,17 @@
 .block-editor-block-list__block {
     position: relative;
 
+    .wp-block {
+
+        &.is-selected {
+            padding: 1.5rem;
+            border-style: solid;
+            border-width: thin;
+            border-radius: 0.25rem;
+            border-color: var(--wp-admin-theme-color);
+        }
+    }
+
```
@glaubersilva did someone mention this as feedback? Let's create a separate PR for this as it affects the form builder UI pretty drastically

glaubersilva replied:
Yes! @JasonTheAdams ask me to add this focus ring in the blocks, so this change makes this happen:

![image](https://github.com/impress-org/givewp-next-gen/assets/16308864/8fac20ca-2be4-401e-963e-23abcf146ce4)

@jdghinson Also did review this focus ring. 

If you notice, I used the same style used in the section border:

![image](https://github.com/impress-org/givewp-next-gen/assets/16308864/51c1511b-470e-4202-a571-31da2fe47b65)

So, I'm not sure about how it can affect the UI drastically, maybe I'm missing something.

jonwaldstein replied:
okay got it, with this addition the padding required causes the fields to jump in place when selected which will affect all blocks.  I'm inclined to move this to a new PR so we can examine its change to the UI closer.

glaubersilva replied:
Resolved in b22dc6d9b04a76b65294e387987b33f71aa76981

New PR here: https://github.com/impress-org/givewp-next-gen/pull/211

jonwaldstein commented on src/DonationForms/resources/app/utilities/BillingAddressStateSelector.tsx:
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+
+declare global {
```
We can remove this file right?

glaubersilva replied:
Resolved in d02f25bdff2aa059674415ad17c30bdad9657062

jonwaldstein commented on src/FormBuilder/resources/js/form-builder/src/blocks/fields/login/icon.tsx:
```diff
@@ -1,8 +1,18 @@
-        <path d="M9.29289 5.29302C9.68342 4.90249 10.3166 4.90249 10.7071 5.29302L14.7071 9.29302C15.0976 9.68354 15.0976 10.3167 14.7071 10.7072L10.7071 14.7072C10.3166 15.0978 9.68342 15.0978 9.29289 14.7072C8.90237 14.3167 8.90237 13.6835 9.29289 13.293L11.5858 11.0001H1C0.447715 11.0001 0 10.5524 0 10.0001C0 9.44784 0.447715 9.00012 1 9.00012H11.5858L9.29289 6.70723C8.90237 6.3167 8.90237 5.6 ...
-    </svg>
-} />
+export default () => (
```
Was the login icon updated in this PR?

glaubersilva replied:
Yes! @jdghinson provided a new icon for the billing address block and ask to also update this one together.

jonwaldstein commented on src/DonationForms/FormDesigns/ClassicFormDesign/css/_inputs.scss:
```diff
@@ -31,7 +31,7 @@ select {
     font-weight: 500;
     line-height: 1.2;
-    padding: 1.1875rem;
+    padding: 0.9rem;
```
why was this updated?

jonwaldstein reviewed (changes requested):
@glaubersilva this is looking great! just a few comments/questions

glaubersilva commented:
@jonwaldstein I did the last change you request and open a new PR for the focus ring change. Let me know if there is anything more.

jonwaldstein reviewed (approved):
@glaubersilva great job man!

rickalday reviewed (approved):
Passed manual QA tests
